File: algorithms/059_spiral_matrix_II.cpp
class Solution {
public:
    vector<vector<int>> generateMatrix(int n) {
        if (n<=0) {
            vector<vector<int> > matrix;
            return matrix;}
        vector<vector<int> > matrix(n,vector<int>(n,0));
        int row=0, col=0,count=0,limit=n*n;
        int direction=1;
        int i=0,j=-1;
        while (count<limit) {
            switch (direction) {
                case 1:
                if (matrix[0].size()-col-1-j>0) j++;
                else {
                    direction =2;
                    i++;
                    row++;}
                break;
                case 2:
                if (i<matrix.size()-row) i++;
                else {direction =3;
                        j--;
                        col++;}
                        break;
                break;
                case 3:
                if (j>=col) j--;
                else {direction =4;
                        i--;}
                        break;
                case 4:
                if (i>row) i--;
                else {
                    direction =1;
                    j++;}
                    break;
            }
            matrix[i][j]=++count;
                }
            return matrix;
        }
};

File: algorithms/python3/059_spiral_matrix_ii.py
from typing import List
class Solution:
    def generateMatrix(self, n: int) -> List[List[int]]:
        seen=[[False]*n for _ in range(n)]
        dr=[0,1,0,-1]
        dc=[1,0,-1,0]
        c=r=di=0
        ans=[[None]*n for _ in range(n)]
        for i in range(n*n):
            ans[r][c]=i+1
            cr,cc=r+dr[di],c+dc[di]
            seen[r][c]=True
            if 0<=cr<n and 0<=cc<n and not seen[cr][cc]:
                r,c=cr,cc
            else:
                di=(di+1) % 4
                r,c=r+dr[di],c+dc[di]
        return ans

a=Solution()

print(a.generateMatrix(4))
